toolchain/check/testdata/var/no_prelude/var_pattern.carbon

```diff
@@ -0,0 +1,330 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// AUTOUPDATE
+// TIP: To test this file alone, run:
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/check/testdata/var/no_prelude/var_pattern.carbon
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/var/no_prelude/var_pattern.carbon
+
+// --- basic.carbon
+
+library "[[@TEST_NAME]]";
+
+fn F() {
+  let var x: () = ();
+}
+
+// --- tuple.carbon
+
+library "[[@TEST_NAME]]";
+
+fn F() {
+  let (x: (), var y: ()) = ((), ());
+}
+
+// --- function.carbon
+
+library "[[@TEST_NAME]]";
+
+fn F(x: (), var y: ());
+
+fn G() {
+  var v: () = ();
+  F(v, ());
+}
+
+// --- fail_nested.carbon
+
+library "[[@TEST_NAME]]";
+
+fn F() {
+  // CHECK:STDERR: fail_nested.carbon:[[@LINE+4]]:27: error: `var` nested within another `var` [NestedVar]
+  // CHECK:STDERR:   let (x: (), var (y: (), var z: ())) = ((), ((), ()));
+  // CHECK:STDERR:                           ^~~
+  // CHECK:STDERR:
+  let (x: (), var (y: (), var z: ())) = ((), ((), ()));
+}
+
+// --- fail_compile_time.carbon
+
+library "[[@TEST_NAME]]";
+
+fn f() {
+  // CHECK:STDERR: fail_compile_time.carbon:[[@LINE+8]]:7: error: semantics TODO: `handle invalid parse trees in `check`` [SemanticsTodo]
+  // CHECK:STDERR:   var T:! type;
+  // CHECK:STDERR:       ^~~~~~~~
+  // CHECK:STDERR:
+  // CHECK:STDERR: fail_compile_time.carbon:[[@LINE+4]]:15: error: `var` declaration cannot declare a compile-time binding [CompileTimeBindingInVarDecl]
+  // CHECK:STDERR:   var T:! type;
+  // CHECK:STDERR:               ^
+  // CHECK:STDERR:
+  var T:! type;
+  let (x: (), T:! type) = ((), ());
+}
+
+// --- fail_implicit.carbon
+
+library "[[@TEST_NAME]]";
+
+// CHECK:STDERR: fail_implicit.carbon:[[@LINE+8]]:10: error: implicit parameters of functions must be constant or `self` [ImplictParamMustBeConstant]
+// CHECK:STDERR: fn F[var u: ()]();
+// CHECK:STDERR:          ^~~~~
+// CHECK:STDERR:
+// CHECK:STDERR: fail_implicit.carbon:[[@LINE+4]]:6: error: found `var` pattern in implicit parameter list [VarInImplicitParams]
```

Comment:
I think we will want to allow `fn F[var self: Self]()` eventually.

Might be worth adding a test for that with a TODO?

Reply:
Oh, nice! I assume we get the "`var` declaration cannot declare a compile-time binding" error for `[var N:! i32]` or similar? Do we have a test for that? (Aside: I suppose that error should say "`var` pattern" rather than "`var` declaration" now.)